lib/rex/proto/ntlm/crypt.rb

```diff
@@ -82,10 +82,11 @@ def self.des_56_to_64(ckey56s)
 
   def self.apply_des(plain, keys)
     raise RuntimeError, "No OpenSSL support" if not @@loaded_openssl
-    dec = OpenSSL::Cipher::DES.new
+    dec = OpenSSL::Cipher.new('DES')
     keys.map do |k|
+      dec.decrypt
       dec.key = k
-      dec.encrypt.update(plain)
+      dec.update(plain)
```

Comment:
This appears to flip the usage from encrypt to decrypt here. Based on the input being "plain", I think it means to encrypt the data.

Reply:
I'm pretty certain this whole block was broken prior to this and is still broken.

TLDR; this should be:
```suggestion
      dec.encrypt
      dec.key = k
      dec.update(plain) + dec.final
```

First, on inspecting `#lm_hash` form IRB, it's not generating a hash.

```
>> Rex::Proto::NTLM::Crypt.lm_hash('test')
=> ""
```

Now, I'm pretty confident that `#apply_des` is returning an array of two empty strings.
```
>> keys = Rex::Proto::NTLM::Crypt.gen_keys('TEST'.ljust(14, "\0"))
=> ["T\"Tj@\x00\x00\x00", "\x00\x00\x00\x00\x00\x00\x00\x00"]
>> Rex::Proto::NTLM::Crypt.apply_des(Rex::Proto::NTLM::Crypt::CONST::LM_MAGIC, keys)
=> ["", ""]
```

This is probably because `dec.update` returns an empty string. That's because after the cipher (en|de)crypt operation, the `#final` method needs to be called, making this `dec.update(plain) + dec.final`. Upon testing this however, I receive a ```in `final': bad decrypt (OpenSSL::Cipher::CipherError)``` exception. After changing it back to `encrypt` though, I see the value I would [expect](https://asecuritysite.com/encryption/lmhash).

```
>> Rex::Proto::NTLM::Crypt.lm_hash('test')
=> "\x01\xFCZk\xE7\xBCi)\x03\xBE\xC6\xE2\xE0\xDCX\f\xAA\xD3\xB45\xB5\x14\x04\xEEG\xD6\x84\xD8\xC0\xF0\xDE\xA9"
```

Reply:
Thanks for catching this! I took a look at this code and started writing tests for it. I then came to the realisation that the code actually exists in [rubyntlm](https://github.com/WinRb/rubyntlm) too. I'm not sure what the historical context is here, or which came first 😄 

Interestingly I also spotted the same fix that you're suggesting in the rubyntlm library too:
https://github.com/WinRb/rubyntlm/commit/f04cff7c6499217b99f3d002f90c1a3aaec25969

As a sanity check, I pointed the rubyntlm's [tests](https://github.com/WinRb/rubyntlm/blob/59879617d14183bca538696722844dc582b4101f/spec/lib/net/ntlm_spec.rb) against crypt.rb and it spotted a different bug with the current implementation. This function doesn't appear to exist on `Rex::Proto::NTLM::Base`, but we do have a Rex::Text equivalent
https://github.com/rapid7/metasploit-framework/blob/19bc85fa1d418c90bf43f8f203f39d33bfdd2430/lib/rex/proto/ntlm/crypt.rb#L132

Our `gen_keys` implementation is also quite different too, and currently fails with rubyntlm's specs

I can put up a separate pr to fix all those issues and potentially reuse the same tests from the ntlm library to verify its behavior, or I could simply keep the same API and delegate responsibility to that gem - as we're already including it:
https://github.com/rapid7/metasploit-framework/blob/19bc85fa1d418c90bf43f8f203f39d33bfdd2430/metasploit-framework.gemspec#L80-L81

Any strong preference? 😄

Reply:
Not a strong preference but delegating the responsibility to reduce duplicate code sounds like a good option.